v2/workloadapi/bundlesource.go

```diff
@@ -132,31 +158,22 @@ func (s *BundleSource) setJWTBundles(bundles *jwtbundle.Set) {
 
 	newBundles := bundles.Bundles()
 
-	// Add/replace the X.509 content in the SPIFFE bundles. Track the trust
-	// domains represented in the new X.509 context so we can determine if
-	// which existing bundles need their X.509 content removed.
+	// Add/replace the JWT authorities from the JWT bundles. Track the trust
+	// domains represented in the new JWT bundles so we can determine which
+	// existing trust domains are no longer represented.
 	trustDomains := make(map[spiffeid.TrustDomain]struct{}, len(newBundles))
 	for _, newBundle := range newBundles {
 		trustDomains[newBundle.TrustDomain()] = struct{}{}
-		existingBundle, ok := s.bundles.Get(newBundle.TrustDomain())
-		if !ok {
-			s.bundles.Add(spiffebundle.FromJWTBundle(newBundle))
-			continue
-		}
-		existingBundle.SetJWTAuthorities(newBundle.JWTAuthorities())
+		s.jwtAuthorities[newBundle.TrustDomain()] = newBundle.JWTAuthorities()
 	}
 
-	// Remove the X.509 content from bundles that are no longer returned
-	// with the X.509 context. If the bundle is then empty, remove the bundle
-	// from the set.
-	for _, existingBundle := range s.bundles.Bundles() {
-		if _, ok := trustDomains[existingBundle.TrustDomain()]; ok {
+	// Remove the JWT authority entries for trust domains no longer represented
+	// in the JWT bundles.
+	for existingTD := range s.jwtAuthorities {
+		if _, ok := trustDomains[existingTD]; ok {
 			continue
 		}
-		existingBundle.SetJWTAuthorities(nil)
-		if existingBundle.Empty() {
-			s.bundles.Remove(existingBundle.TrustDomain())
-		}
+		delete(s.jwtAuthorities, existingTD)
```

Comment:
This should be protected by a lock.

Reply:
It is. See the top of the setJWTBundles function.